### PR TITLE
fix: use map to store policy index

### PR DESCRIPTION
**Pull request by dovics**
fix: #499

### CODE REVIEW

**Comment by hsluoyz** [comment]
@nodece @GopherJ @UnderTreeTech please review.

File: model/policy.go [hunk 1]
```diff
@@ -123,34 +124,37 @@ func (model Model) HasPolicies(sec string, ptype string, rules [][]string) bool
...
 
 // RemovePolicies removes policy rules from the model.
 func (model Model) RemovePolicies(sec string, ptype string, rules [][]string) bool {
-	for j := 0; j < len(rules); j++ {
-		for i, r := range model[sec][ptype].Policy {
-			if util.ArrayEquals(rules[j], r) {
-				model[sec][ptype].Policy = append(model[sec][ptype].Policy[:i], model[sec][ptype].Policy[i+1:]...)
-			}
+	for _, rule := range rules {
+		if ok := model.RemovePolicy(sec, ptype, rule); !ok {
```
Comment (nodece) [inline model/policy.go:156]: I think the return value should be ignored.

**Comment by nodece** [comment]
Hi @dovics , Could you add benchmark test for the changes?  I want to know if the performance has improved.

File: model/policy.go [hunk 2]
```diff
@@ -123,34 +124,37 @@ func (model Model) HasPolicies(sec string, ptype string, rules [][]string) bool
...
-			return true
-		}
+	index, ok := model[sec][ptype].PolicyMap[strings.Join(rule, DefaultSep)]
+	if !ok {
+		return false
 	}
 
-	return false
+	model[sec][ptype].Policy = append(model[sec][ptype].Policy[:index], model[sec][ptype].Policy[index+1:]...)
+	delete(model[sec][ptype].PolicyMap, strings.Join(rule, DefaultSep))
```
Comment (UnderTreeTech) [inline model/policy.go:145]: concurrent delete map element will panic

**Comment by UnderTreeTech** [comment]
It's exactly a great improvement, thanks for your work.

File: model/policy.go (hunk 2 above)
Comment (nodece) [reply model/policy.go:145]: Caspin should use thread safe instance by default.

**Comment by dovics** [comment]
The Benchmark previous
Screen Shot 2020-06-20 at 12 25 43 PM

The Benchmark later
Screen Shot 2020-06-20 at 12 28 20 PM

**Comment by UnderTreeTech** [comment]
> The Benchmark previous
> Screen Shot 2020-06-20 at 12 25 43 PM
> 
> The Benchmark later
> Screen Shot 2020-06-20 at 12 28 20 PM

Exciting!

File: model/policy.go (hunk 2 above)
Comment (dovics) [reply model/policy.go:145]: I think the issue of thread safety should be provided by SyncedEnforcer. Before, the model was not thread safe. Even if we guarantee that map is thread-safe, policy slice still requires addition another lock

File: model/assertion.go [hunk 3]
```diff
@@ -25,11 +25,12 @@ import (
...
-	Value  string
-	Tokens []string
-	Policy [][]string
-	RM     rbac.RoleManager
+	Key       string
+	Value     string
+	Tokens    []string
+	Policy    [][]string
+	PolicyMap map[string]int
+	RM        rbac.RoleManager
```
Comment (nodece) [inline model/assertion.go:33]: Could you add `sync.RWMutex` to the Assertion?

File: model/policy.go (hunk 2 above)
Comment (nodece) [reply model/policy.go:145]: @dovics you're right.

File: model/assertion.go (hunk 3 above)
Comment (dovics) [reply model/assertion.go:33]: ok, I will add it

Comment (dovics) [reply model/assertion.go:34]: I still think this mutex has some duplicates with SyncedEnforce

Comment (nodece) [reply model/assertion.go:34]: I see, we can new a  PR to improve it.

File: model/policy.go [hunk 4]
```diff
@@ -122,46 +140,54 @@ func (model Model) HasPolicies(sec string, ptype string, rules [][]string) bool
...
+	defer model[sec][ptype].Mutex.Unlock()
 	model[sec][ptype].Policy = append(model[sec][ptype].Policy, rule)
+	model[sec][ptype].PolicyMap[strings.Join(rule, DefaultSep)] = len(model[sec][ptype].Policy) - 1
 }
 
 // AddPolicies adds policy rules to the model.
 func (model Model) AddPolicies(sec string, ptype string, rules [][]string) {
-	for i := 0; i < len(rules); i++ {
-		model[sec][ptype].Policy = append(model[sec][ptype].Policy, rules[i])
+	for _, rule := range rules {
```
Comment (nodece) [inline model/policy.go:151]: Add lock here.

**Comment by nodece** [comment]
@dovics Did you miss the https://github.com/casbin/casbin/pull/500/files#diff-abe2d810a9c1f824dbc2aa7b8e3afd26R252?

**Pull request merged**